content/part4.md

```diff
@@ -0,0 +1,199 @@
+
+## Handler Functions
+
+First, let's take a look at the `POST /cart` endpoint. Since we've chosen to represent an individual shopping cart as a Workflow, the `CreateCartHandler()` function will create a new Workflow using `ExecuteWorkflow()`.
+You are responsible for making sure each `POST /cart` call creates a Workflow creates a unique `workflowID`.
+[Temporal treats a Workflow with the same ID as a prior Workflow as a continuation of the prior Workflow](https://docs.temporal.io/docs/shared/continue-as-new).
+
+```go
+func CreateCartHandler(w http.ResponseWriter, r *http.Request) {
+  // In production you should use uuids or something similar, but the
+  // current time is enough for this example. Make sure the Workflow ID
+  // is unique every time the user creates a new cart!
+	workflowID := "CART-" + fmt.Sprintf("%d", time.Now().Unix())
+
+	options := client.StartWorkflowOptions{
+		ID:        workflowID,
+		TaskQueue: "CART_TASK_QUEUE",
+	}
+
+	cart := app.CartState{Items: make([]app.CartItem, 0)}
+	we, err := temporal.ExecuteWorkflow(context.Background(), options, app.CartWorkflow, cart)
+	if err != nil {
+		WriteError(w, err)
+		return
+	}
+
+  // Return the `workflowID` and `runID` so clients can use them with other endpoints
+	res := make(map[string]interface{})
+	res["cart"] = cart
+	res["workflowID"] = we.GetID()
+	res["runID"] = we.GetRunID()
+
+	w.WriteHeader(http.StatusCreated)
+	json.NewEncoder(w).Encode(res)
+}
+```
+
+Now you have a `POST /cart` endpoint that creates a new empty cart, and returns the `workflowID` and `runID` that uniquely identify this Workflow.
+Even though `workflowID` is unique in this case, you still need to provide `runID` in order to Query or Signal the cart Workflow.
+The next endpoint is `GET /cart/{workflowID}/{runID}`, which returns the current state of the cart with the given `WorkflowID` and `runID`.
+Below is the `GetCartHandler()` function, which gets the `workflowID` and `runID` from the URL and executes a Query for the current state of the cart.
+
+```go
+func GetCartHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	response, err := temporal.QueryWorkflow(context.Background(), vars["workflowID"], vars["runID"], "getCart")
+	if err != nil {
+		WriteError(w, err)
+		return
+	}
+	var res interface{}
+	if err := response.Get(&res); err != nil {
+		WriteError(w, err)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(res)
+}
+```
+
+## PUT Requests and Signals
+
+HTTP PUT requests usually correspond to Temporal Signals.
+That means, in addition to the `workflowID` and `runID`, you need to send Signal arguments.
+Remember that `shared.go` contains an [`AddToCartSignal` struct](https://github.com/temporalio/temporal-ecommerce/blob/5c4e0142e3571398d972c80b3fa7cdbe7a5db42b/shared.go#L64-L67) that is what the [cart Workflow's Signal handler expects](https://github.com/temporalio/temporal-ecommerce/blob/main/workflow.go#L52-L71):
```

Comment:
```suggestion
Remember that `shared.go` contains an [`AddToCartSignal` struct](https://github.com/temporalio/temporal-ecommerce/blob/5c4e0142e3571398d972c80b3fa7cdbe7a5db42b/shared.go#L64-L67) which is what the [cart Workflow's Signal handler expects](https://github.com/temporalio/temporal-ecommerce/blob/main/workflow.go#L52-L71):
```

Reply:
Also I would say that "HTTP PUT requests usually correspond to Temporal Signals." makes it seem like an official rule rather than an arbitrary (but IMO good choice) mapping you've made in the design.